shell/platform/linux/fl_platform_plugin.cc

```diff
@@ -140,6 +150,45 @@ static FlMethodResponse* clipboard_has_strings_async(
   return nullptr;
 }
 
+// Called when Flutter wants to exit the application.
+static FlMethodResponse* system_exit_application(FlPlatformPlugin* self,
+                                                 FlValue* args) {
+  if (fl_value_get_type(args) != FL_VALUE_TYPE_MAP) {
+    return FL_METHOD_RESPONSE(fl_method_error_response_new(
+        kBadArgumentsError, "Argument map missing or malformed", nullptr));
+  }
+
+  FlValue* type_value = fl_value_lookup_string(args, kExitTypeKey);
+  if (type_value == nullptr ||
+      fl_value_get_type(type_value) != FL_VALUE_TYPE_STRING) {
+    return FL_METHOD_RESPONSE(fl_method_error_response_new(
+        kBadArgumentsError, "Missing exit type", nullptr));
+  }
+  const gchar* type = fl_value_get_string(type_value);
+  bool is_cancelable = g_str_equal(type, kExitTypeCancelable);
+
+  bool canceled = false;
+  if (is_cancelable) {
+    // TODO
```

Comment:
For cancelable requests, the TODO code here will need to send a request back to the framework so that any registered listeners for exit events get a chance to respond to the request, so the rest of this function will need to be inside the "reply" callback for that method channel call. 

This bouncing back into the framework is so that if, for instance, a package registers, it can get notified, and so that requests from the OS (like on session end) go through the same path as the framework-originated requests (which should also go through the method call).

I could also short circuit the request in the framework and only tell the engine to exit when it is required to, but I figured fewer different paths was better. And we'd still need almost the same code to handle requests from the OS.

Reply:
Oh, I thought the handler for this was in the native code. I'll update this to the desired behaviour.

Reply:
Should be good now, please check again!